pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go

```diff
@@ -494,56 +494,70 @@ func revertToCutoverTimestamp(
-	db := p.ExecCfg().DB
-	jobRegistry := p.ExecCfg().JobRegistry
-	ingestionJob, err := jobRegistry.LoadJob(ctx, ingestionJobID)
-	if err != nil {
-		return false, err
-	}
-
-	var shouldRevertToCutover bool
-	if err := p.ExecCfg().InternalDB.Txn(ctx, func(
-		ctx context.Context, txn isql.Txn,
-	) error {
-		return ingestionJob.WithTxn(txn).Update(ctx, func(txn isql.Txn, md jobs.JobMetadata, ju *jobs.JobUpdater) error {
-			payload := md.Payload.GetStreamIngestion()
-			if payload == nil {
-				return errors.Newf("unknown payload type %T in stream ingestion job %d",
-					md.Payload, ingestionJobID)
-			}
-			streamIngest := md.Progress.GetStreamIngest()
-			if streamIngest == nil {
-				return errors.Newf("unknown progress type %T in stream ingestion job %d",
-					md.Progress, ingestionJobID)
-			}
-			cutoverTime := streamIngest.CutoverTime
-			if cutoverTime.IsEmpty() {
-				log.Infof(ctx, "empty cutover time, no revert required")
-				return nil
+	// The update below sets the ReplicationStatus to
+	// CuttingOver. Once set, the cutoverTimestamp cannot be
+	// changed. We want to be sure to read the timestamp that
+	// existed in the record at the point of the update rather the
+	// value that may be in the job record before the update.
+	var (
+		shouldRevertToCutover bool
+		cutoverTimestamp      hlc.Timestamp
+		spanToRevert          roachpb.Span
+	)
+	if err := ingestionJob.NoTxn().Update(ctx,
+		func(txn isql.Txn, md jobs.JobMetadata, ju *jobs.JobUpdater) error {
+			streamIngestionDetails := md.Payload.GetStreamIngestion()
+			if streamIngestionDetails == nil {
+				return errors.AssertionFailedf("unknown payload %v in stream ingestion job %d",
+					md.Payload, ingestionJob.ID())
 			}
-			if md.Progress.GetHighWater() == nil || md.Progress.GetHighWater().Less(cutoverTime) {
-				log.Infof(ctx, "job with highwater %s not yet ready to revert to cutover at %s",
-					md.Progress.GetHighWater(), cutoverTime.String())
-				return nil
+
+			streamIngestionProgress := md.Progress.GetStreamIngest()
+			if streamIngestionProgress == nil {
+				return errors.AssertionFailedf("unknown progress %v in stream ingestion job %d",
+					md.Progress, ingestionJob.ID())
 			}
 
-			shouldRevertToCutover = true
-			updateRunningStatusInternal(md, ju, jobspb.ReplicationCuttingOver,
-				fmt.Sprintf("starting to cut over to the given timestamp %s", cutoverTime))
+			cutoverTimestamp = streamIngestionProgress.CutoverTime
+			spanToRevert = streamIngestionDetails.Span
+			shouldRevertToCutover = cutoverTimeIsEligibleForCutover(ctx, cutoverTimestamp, md.Progress)
+
+			if shouldRevertToCutover {
+				updateRunningStatusInternal(md, ju, jobspb.ReplicationCuttingOver,
+					fmt.Sprintf("starting to cut over to the given timestamp %s", cutoverTimestamp))
+			} else {
+				if streamIngestionProgress.ReplicationStatus == jobspb.ReplicationCuttingOver {
+					return errors.AssertionFailedf("cutover already started but cutover time %s is not eligible for cutover",
```

Comment:
Let's visit in a follow up. We shouldn't need to in the current structure since the frontier can only move forward (unless there is a bug) and the frontier being _ahead_ of the cutover time is fine.